Device Plugins Framework Refactor | Part 2

Aseeef opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does

This PR refactors and improves the device plugin health check logic.

*   Device plugins now default to an `Unhealthy` state until a successful health check is performed.
*   `mediated_device.go`, `pci_device.go`, and `socket_device.go` add the parent of the parent directory add a fallback in case the device directory is not there.  Motivation explained in references.
*   Previously, some devices were watching files instead of directories. fsnotify documentations warn against this because once the file is deleted, the watcher on it is lost. Therefore, we improve on this behavior by having fsnotify only watch directories.
*   Added callbacks like `SetupMonitoredDevices` and `ConfigurePermissions` to allow device-specific logic for health checking and permission management.
*   Fixed a bug in `usb_device`.  `usb_device` reports health on a group-level (i.e. a group of USB devices share the same health). Logically, then what *should* happen is that if any device in the group is unhealthy, the entire group is unhealthy. In practice, what *was* happening is the last device in the group to report its health decided the health for the entire group.
*   Fixed a potential resource leak scenario with the healthCheck routine where if the device plugin stopped right before the healthCheck loop sends out a health update, it would block forever since nothing was listening on the other side
*   More uniform, cleaner, and detailed logging messages.
*   Expanded and improved test cases, and updated test cases to now expect devices to start out as unhealthy. `[*]`

#### Before this PR:

Health check logic between the device plugins had significant duplication and inconsistent behavior.

#### After this PR:

All duplicate logic is consolidated. In theory, someone implementing a new device plugin no longer has to understand the full health check logic due to how the logic was now centralized.

### References
Refactor effort tracker: https://github.com/kubevirt/kubevirt/issues/17677

PR 1: https://github.com/kubevirt/kubevirt/pull/17666
PR 2: This
PR 3: https://github.com/kubevirt/kubevirt/pull/17676

### Why we need it and why it was done in this way

The following tradeoffs were made:
* `ConfigurePermissions` introduces a behavior change particularly for `socket_device.go`. While most of this refactor attempts to maintain the same functional behavior as the non-refactored code, `ConfigurePermission` is an exception. While previously, the code that is now included in `ConfigurePermission` was in the constructor for `NewSocketDevicePlugin`, `ConfigurePermission` is now run exclusively in the `healthCheck` logic and removed from the constructor because I as far as I can tell running this permission setting logic slightly later has no harm and helps centralize the lifecycle logic for when permissions are set. (`usb_device.go` also makes permission changes in Allocate but I thought it best to leave that one alone).
  * Adding permission configuration into `healthCheck` was seen as necessary to me to address the following situation: device exists ⇾ plugin device is marked healthy ⇾ device gets allocated to a VM ⇾ device gets deleted (for whatever reason) ⇾ device is added back. Since in this old behavior, `Allocate()` is never called again, permissions for this new device would never get configured even if the device is healthy again.

### Special notes for your reviewer
* I had Gemini 3.0 review my PR, and it pointed out there could be a race conditions with plugin devices where say usb_device and udev rules both try to make permission changes at the same time if a USB device is removed and re-added (hot plugging). Technically, this issue isn't directly caused by this PR, and fixing this would significantly complicate this already complicated PR further. I think this should be dealt with in a separate PR.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

0xFelix commented on pkg/virt-handler/device-manager/generic_device_test.go:
```diff
@@ -82,8 +88,14 @@ var _ = Describe("Generic Device", func() {
 
 		os.OpenFile(dpi.socketPath, os.O_RDONLY|os.O_CREATE, 0666)
 
-		go dpi.healthCheck()
```
No need to start healthCheck anymore? In the test above it's still started?

0xFelix commented on pkg/virt-handler/device-manager/mediated_device.go:
```diff
@@ -126,7 +76,8 @@ func NewMediatedDevicePlugin(mdevs []*MDEV, resourceName string) *MediatedDevice
 	}
-
+	dpi.SetupMonitoredDevices = dpi.SetupMonitoredDevicesFunc
+	dpi.GetIDDeviceName = dpi.GetIDDeviceNameFunc
```
Same as above

0xFelix commented on pkg/virt-handler/device-manager/pci_device.go:
```diff
@@ -116,15 +66,17 @@ func NewPCIDevicePlugin(pciDevices []*PCIDevice, resourceName string) *PCIDevice
 		iommuToPCIMap: iommuToPCIMap,
 	}
+	dpi.SetupMonitoredDevices = dpi.SetupMonitoredDevicesFunc
+	dpi.GetIDDeviceName = dpi.GetIDDeviceNameFunc
```
Same

0xFelix commented on pkg/virt-handler/device-manager/pci_device.go:
```diff
@@ -175,87 +127,38 @@ func (dpi *PCIDevicePlugin) Allocate(_ context.Context, r *pluginapi.AllocateReq
 	err := filepath.Walk(pciBasePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
```
Ouch, thanks!

0xFelix commented on pkg/virt-handler/device-manager/socket_device.go:
```diff
@@ -129,96 +59,85 @@ func (dpi *SocketDevicePlugin) setSocketPermissions() error {
-	if err != nil {
-		return err
+	dpi.SetupMonitoredDevices = dpi.SetupMonitoredDevicesFunc
+	dpi.GetIDDeviceName = dpi.GetIDDeviceNameFunc
```
Same

0xFelix commented on pkg/virt-handler/device-manager/usb_device_test.go:
```diff
@@ -77,7 +77,6 @@ var _ = Describe("USB Device", func() {
 		for resourceName, pluginDevices := range pdmap {
 			Expect(pluginDevices).To(HaveLen(len(result[resourceName])), "Number of k8s devices")
 			for i, dev := range pluginDevices {
-				Expect(dev.isHealthy).To(BeTrue())
```
Check transition from unhealthy to healthy too?

0xFelix commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -36,26 +37,100 @@ import (
+	}()
+
+	err = waitForGRPCServer(dpi.socketPath, connectionTimeout)
+	if err != nil {
```
Same

0xFelix commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -36,26 +37,100 @@ import (
+	}
+
+	err = dpi.register()
+	if err != nil {
```
Same

0xFelix reviewed (commented):
Thanks, this is awesome! Added some comments.

Did you try enabling the linted for this path? `pkg/virt-hanlder-device-manager`

Aseeef commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -131,17 +259,45 @@ func (dpi *DevicePluginBase) healthCheck() error {
+				// If event for the parent directory add a watcher for the device directory
+				// This code path can only be triggered if parent directory was added to the watcher.
+				logger.Infof("device directory %s was created, adding watcher", deviceDirPath)
+				if err := watcher.Add(deviceDirPath); err != nil {
```
When the health check exits, yes. There is a defer at the start of the healthCheck function.

Aseeef commented:
@0xFelix I think I have addressed all the mentioned concerns except for how `usb_device` (and actually for that matter even `pci_device` and `mediated_device`) don't have good test coverage. I can definitely look into addressing these concerns and fully battle testing my changes as well.

0xFelix commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -36,26 +37,100 @@ import (
+	initialized           bool
+	lock                  *sync.Mutex
+	deregistered          chan struct{}
+	deviceRoot            string                                           // Root directory where the device is located
```
ACK

0xFelix commented on pkg/virt-handler/device-manager/generic_device.go:
```diff
@@ -21,223 +21,75 @@ package device_manager
 
 	for i := 0; i < maxDevices; i++ {
-		deviceId := dpi.deviceName + strconv.Itoa(i)
+		deviceId := deviceName + strconv.Itoa(i)
```
nit: You could use `for i := range maxDevices` above

0xFelix commented on pkg/virt-handler/device-manager/generic_device.go:
```diff
@@ -21,223 +21,75 @@ package device_manager
 		if err == nil {
 			devnode.Close()
+		} else {
+			return fmt.Errorf("error opening the device node %s: %v", devicePath, err)
```
Also stylistic nit: Could be an early return and then just call `devnode.Close()`

```suggestion
		if err != nil {
			return fmt.Errorf("error opening the device node %s: %v", devicePath, err)
		}
		devnode.Close()
```

0xFelix commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -36,26 +37,100 @@ import (
+	// for additional setup steps that are not covered by the default implementation
+	if dpi.SetupDevicePlugin != nil {
+		err = dpi.SetupDevicePlugin()
+		if err != nil {
```
Stylistic nit about inlining

0xFelix reviewed (commented):
Do you think you could enhance the unit test suite first before merging this PR? This way we could show that the refactoring does not introduce any regressions.

Aseeef commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -88,41 +164,93 @@ func (dpi *DevicePluginBase) ListAndWatch(_ *pluginapi.Empty, s pluginapi.Device
 
 func (dpi *DevicePluginBase) healthCheck() error {
 	logger := log.DefaultLogger()
+
```
I dont think I understand what you mean by this

Aseeef commented on pkg/virt-handler/device-manager/socket_device.go:
```diff
@@ -129,96 +59,82 @@ func (dpi *SocketDevicePlugin) setSocketPermissions() error {
 }
 
-func NewSocketDevicePlugin(socketName, socketDir, socket string, maxDevices int, executor selinux.Executor, p PermissionManager) (*SocketDevicePlugin, error) {
+func NewSocketDevicePlugin(socketName, socketDir, socketFile string, maxDevices int, executor selinux.Executor, p PermissionManager) (*SocketDevicePlugin, error) {
```
This can now just return (*SocketDevicePlugin) since this function can no longer produce an error. We will always succeed in creating the device plugin. It might be unhealthy, but now by design it can recover.

Aseeef commented:
> PR needs rebase.

Will resolve Monday. In the meantime, could I have some guidance in regard to the release note? Release notes are user facing and technically from the user POV, device plugins are an implementation detail users don't care about. At the same time, there are some bug fixes, logging changes and consistency improvements in this PR so I don't think I would put "NONE"..?

vladikr commented on pkg/virt-handler/device-manager/generic_device.go:
```diff
@@ -21,223 +21,74 @@ package device_manager
-	_, err = client.Register(context.Background(), reqt)
-	if err != nil {
-		return err
+		defer devnode.Close()
```
This should be before the return, otherwise it's unreachable...

vladikr commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -64,7 +139,8 @@ func (dpi *DevicePluginBase) ListAndWatch(_ *pluginapi.Empty, s pluginapi.Device
-				if devHealth.DevId == dev.ID {
+				// If the devHealth.DevId is empty, it was not set by the device plugin, so we update all devices
+				if devHealth.DevId == dev.ID || devHealth.DevId == "" {
 					dev.Health = devHealth.Health
```
@Aseeef isn't this a race? looking at this, it seems that healthcheck can update the dpi.health while we;re moifying deviceHealth.Health without a lock...

Aseeef replied:
ListAndWatch is the only place where Health is ever updated as far as I can see.

vladikr replied:
Yea.. ok.

vladikr reviewed (commented):
I didn't look at this code for a while. It changed a lot recently. 
I found some issues...

0xFelix reviewed (commented):
There are some stylistic and repeated errors? Can you please fix them in each file? I did not comment on every single occurence.

zhencliu commented:
/retest-required

zhencliu commented:
/retest-required

Aseeef commented:
/unhold

Aseeef commented:
@0xFelix since you last looked at it, I put all new changes in: https://github.com/kubevirt/kubevirt/pull/16073/changes/3d08c7ebada3fd2eea0e7f7b9c8e2bbffdf9cf2d

* Addressed nits you pointed out.
* Fixed a race condition (this didn't end up being the reason my tests kept failing, the real reason was wayy more trivial and this issue was there before). The race condition was that dpi.healthCheck() is run in a separate routine, and it was possible it could miss events if healthCheck hadn't started, and you modified a file (as with some of the test) though in practice I don't think this ever happened to me.
  * This also split healthCheck up even more so its a smaller function now.
* Allocate() didn't run as expected (turns out when I was testing on my kubevirt instance things were working because code didn't properly get updated).
* Fixed failing e2e due to a log message change.

awels commented:
/unhold
/test pull-kubevirt-e2e-k8s-1.32-sig-compute

enp0s3 reviewed (changes requested):
@Aseeef Hi, thank you for taking care of this!
Can you please split the first commit into multiple commits, each with its own context (grammar fix, naming consistency, introduction of the permission manager) otherwise its very hard to review a dense commit with multiple changes, it creates cognitive load on the reviewer to map each file change to the context of the change.

Aseeef committed:
Fixed grammar and formatting issues in device plugin api.proto

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Moved PermissionManager interface to shared location

PermissionManager interface and relevant methods were moved to common.go because it is a general interface that can be applicable to device plugins outside just socket device. Previously this interface was also unnecessarily exported. This is also no longer the case.

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Centralized device plugin Setup into device_plugin_base.go removing c&p code in other device plugins.

Start(), stopDevicePlugin(), register(), GetInitialized(), setInitialized() were re-implemented or duplicated in each device plugin. Now the base device plugin contains logic for all of these reducing duplication and resulting in uniform behavior across all DPs.

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Moved a shared interface & constants from generic_device to more centralized location

The Device interface is implemented by all device plugins and belongs in a more central file like device_controller since the interface is not specific to generic device. Moreover, device_controller is also the consumer of this interface.

Second, constants were moved to device_plugin_base.

Third, interface fields like ListAndWatch, Allocate, PreStartContainer are redundant and were removed since device controller already enforced that all DPs be of the interface type DevicePluginClient.

Signed-off-by: Aseef <aimran@redhat.com>

Aseeef committed:
Added a deviceNameByID function for retrieving a log-friendly name for a device plugin.

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Moved the Allocate function to device_plugin_base.

Each device plugin implements its own Allocate function by registering a callback function with device_plugin_base.

While technically, you can leave each device plugin to implement its own Allocate() and everything should still work, this is done for consistency and to avoid any footguns with Go's composition model.

The second benefit of doing this is that if someone creates a new device plugin later, the device_plugin_base will throw a runtime error if they forget to provide an Allocate function.

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Updated unit tests & e2e tests

* Unit tests and e2e tests were updated to expect device plugins to start as unhealthy (since previously all DPs started as healthy and the unit tests expected this).
* E2E tests expected a particular log format so updated logging format to match current behavior.
* Added several more unit tests increasing test coverage.
* Many unit tests did not work as intended. For example, generic_device polls dpi.devs[0].Health when in the setup of the tests, this field never gets updated. This behavior was fixed.
* Syntactic improvements

Signed-off-by: Aseef Imran <aimran@redhat.com>
Signed-off-by: Aseef <contact@aseef.dev>

enp0s3 commented on pkg/virt-handler/device-manager/common.go:
```diff
@@ -259,3 +260,17 @@ type deviceHealth struct {
 	DevId  string // Note, when DevId is empty, it means the device is all devices
 	Health string
 }
+
```
@Aseeef Please note that permission manager was accessible to everyone in the device_manager package in the first place, so I am not sure what is the added value of moving it to common.go. Moreover, both the `PermissionManager` interface and the function:
```
func NewPermissionManager() PermissionManager {
	return &permissionManager{}
}
```
Are exported functions (beginning with Capital Letter) so these functions are even accessible outside of the device_manager package, which seems strange to me. 
From refactoring perspective I would recommend to to the following:
1. Unexport the permission manager since I don't see any added value in it
2. Refrain from using go mock, to me it looks like an overkill for the permission manager. Since it is an interface it can be re-implemented in the unit tests w/o any third party generators.

enp0s3 commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -34,25 +36,79 @@ import (
+	socketPath        string
+	stop              <-chan struct{}
+	health            chan deviceHealth
+	resourceName      string // The kubernetes resource name for this device plugin
```
Comments usually indicate a code smell. If you think the variable name is misleading then lets refactor the variable name in a dedicated commit.

enp0s3 commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -34,25 +36,79 @@ import (
+	initialized       bool
+	lock              *sync.Mutex
+	deregistered      chan struct{}
+	deviceRoot        string       // Root directory where the device is located
```
Commit is redundant IMO

enp0s3 reviewed (changes requested):
@Aseeef Thank you for the patience! I want over most of the changes, I would like to post the first iteration of comments so it will be easier to review. 
In general in would also be much easier for me to understand the context of each change if you would follow [the commit message best practices guilde](https://www.berrange.com/tags/commit-message/)

Aseeef committed:
Renamed "GetDeviceName" to "GetResourceName"

Renamed "GetDeviceName" to "GetResourceName" to more accurately reflect what the function does (and avoid confusing terminology overloading with other similar fields). Also removed the "deviceName"/"socketName" field since it was redundant and not used outside logging (and its usage was not even consistent).

Logs also now use "resourceName" instead, and so an E2E test had to be updated.

Signed-off-by: Aseef <contact@aseef.dev>

enp0s3 reviewed (commented):
@Aseeef @zhencliu Hey folks. Why does #16262 depends on this PR?

enp0s3 reviewed (commented):
@Aseeef Thank you for the patience!
I left some comments, after going over them can you please consider the following:
commits:
[d14cb8d71e60ace63f8912d4c923166e1dad8db6](https://github.com/kubevirt/kubevirt/pull/16073/commits/d14cb8d71e60ace63f8912d4c923166e1dad8db6)
[6070259ba26b585a69cd4ff7f98a64e814e8590e](https://github.com/kubevirt/kubevirt/pull/16073/commits/6070259ba26b585a69cd4ff7f98a64e814e8590e)
Consider drop these commits because I think the whole usage of device plugin base in the following manner:
```
type SomeDevicePlugin struct {
	*DevicePluginBase
...
}
```
is a bad design and should be refactored. device plugins have two implementation contracts that they need to satisfy: one with the device controller and another one is with the device plugin gRPC server. 
So I think we need to convert the device plugin base into a helper library package. Device plugins would be able consume the helper package functions or to implement their own methods. 

Regarding commit [14e9888d418c317da7564b73bb1853ec1d64e9e0](https://github.com/kubevirt/kubevirt/pull/16073/commits/14e9888d418c317da7564b73bb1853ec1d64e9e0) you have a lot of bullets there, can you please try to split them to separate commits?

Regarding commit [0a715f11a0b4183716bcde4efa64963fa7089050](https://github.com/kubevirt/kubevirt/pull/16073/commits/0a715f11a0b4183716bcde4efa64963fa7089050) can you please take it into separate PR?

As a reference you can take a look on the following refactoring effort:
https://github.com/kubevirt/kubevirt/issues/16117

You can follow this pattern. When the changes are incremental they have bigger chance to converge faster.

zhencliu commented:
> > > @Aseeef @zhencliu Hey folks. Why does #16262 depends on this PR?
> > 
> > 
> > Hi @enp0s3, thanks for asking. PR #16262 leverages the device plugin mechanism to advertise secure guest capacity resources. Since @Aseeef is currently refactoring the framework, I’ve based my PR on top of that one to stay aligned.
> 
> @zhencliu Thanks for the clarification. If the refactoring doesn't solve any issue for your PR I would recommend to un-rebase because these changes are still under discussion and things may change.

Thanks @enp0s3 . It makes sense to me.

Aseeef committed:
generic_device: explicit device path

Previously, generic_device defined dpi.devicePath as the target file path for the "preOpen" step. This would, for example, map to /dev/kvm.

While (technically) this is correct since virt-handler is a priviledged pod and therefore automatically maps /dev from the host into /dev on the pod, it is more consistent and explicit to use  filepath.Join(dpi.deviceRoot, dpi.devicePath) as the device path.

In short, both work but this is better.

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
USB devices cleanup

* usb_device now uses PermissionManager (like socket_device) to run ChownAtNoFollow for consistency.
* Must now pass in deviceRoot field into USBDevicePlugin constructor. Reason: right now usb_device lacks test. In the future, this field will be useful for unit testing.
* Now setting "devs" field in usb_device so DevicePluginBase is able to see devices usb_device has.

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Socket devices cleanup

* socket_device: Added support for host-mounted device plugins (host-mounted socket device plugins don't currently exist but will be needed soon in future PRs)
* Removed unnecessary fields in socket_device redefining logic primarily in terms of fields available via DevicePluginBase.

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Various cleanups & minor fixes

* dpi.server field removed in mediated_device (already set in device_plugin_base)
* Fixed an uncaught error in pci_device filepath.Walk.

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Plugin devices now start unhealthy until proven otherwise

Previously, device plugins would start as "healthy" (and therefore allocable) even before we confirm whether they are actually healthy.

This allowed for a small window where a potentially unhealthy device could be allocated. Worst, if an error prevents health check loop from starting, a possibly unhealthy device can advertise itself as healthy!

Therefore, a default where all devices begin as unhealthy until we are sure otherwise is more logical.

(Note: this requires adapting tests to this new behavior)

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Refactored & drastically improved healthCheck logic in device plugins

* Health check lifecycle has been centralized in device_plugin_base using callback functions that individual DPs can register.
* Device plugin will now ensure file watchers are setup before marking the DP as "initialized". Since this is done synchronously, the health update channel is now also bounded to prevent blocking the main thread if the Kubelet isn't yet ready to start consuming health updates.
* Various logging improvements
* Device plugins now default to an Unhealthy state until a successful health check is performed.
* mediated_device.go, pci_device.go, and socket_device.go now add the parent of the parent directory add a fallback in case the device directory is not there.
* Previously, some devices were watching files instead of directories. fsnotify documentations warn against this because once the file is deleted, the watcher on it is lost. Therefore, we improve on this behavior by having fsnotify only watch directories.
* healthCheck loop also monitors permission changes and posts them to logs for better visibility of possible failures due to permission issues.
* healthCheck also now tracks the state of the previous device health status to reduce the number of unnecessary updates pushed to consumers.
* Fixed a potential resource leak scenario with the healthCheck routine where if the device plugin stopped right before the healthCheck loop sends out a health update, it would block forever since nothing was listening on the other side
* Overall more robust health check behavior that is resiliant to events like (a) Kubelet temporarily becoming unavailable, (b) Multiple listeners to ListAndWatch

Signed-off-by: Aseef <aimran@redhat.com>
Signed-off-by: Aseef <contact@aseef.dev>

Aseeef committed:
Fixed usb_device health reporting bug

* Fixed a bug in usb_device. usb_device reports health on a group-level (i.e. a group of USB devices share the same health). Logically, then what should happen is that if any device in the group is unhealthy, the entire group is unhealthy. In practice, what was happening is the last device in the group to report its health decided the health for the entire group.
* With the addition of "mutateHealthUpdate" also simplified "OptionalSocketDevice"

Signed-off-by: Aseef <contact@aseef.dev>

Aseeef commented:
/test pull-kubevirt-e2e-k8s-1.34-sig-compute-serial

0xFelix commented on pkg/virt-handler/device-manager/socket_device_test.go:
```diff
@@ -48,12 +48,10 @@ var _ = Describe("Socket device", func() {
 
 		ctrl := gomock.NewController(GinkgoT())
 		mockExec := selinux.NewMockExecutor(ctrl)
-		mockPermManager := NewMockPermissionManager(ctrl)
```
Why drop the mock permission manager? Sounds unsafe to me to use the real one in testing environments.

0xFelix commented on pkg/virt-handler/device-manager/mediated_device.go:
```diff
@@ -315,7 +269,7 @@ func (dpi *MediatedDevicePlugin) healthCheck() error {
 
 					dpi.health <- deviceHealth{
 						DevId:  monDevId,
-						Health: pluginapi.Unhealthy,
```
When a device was removed, shouldn't it be unhealthy?

0xFelix commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -154,6 +177,110 @@ func (dpi *DevicePluginBase) getFriendlyName(deviceID string) string {
+		return healthCtx, fmt.Errorf("setupMonitoredDevices is not implemented")
+	}
+	err = dpi.setupMonitoredDevices(watcher, healthCtx.monitoredDevices)
+	if err != nil {
```
nit: inline

0xFelix commented on tests/vmi_lifecycle_test.go:
```diff
@@ -620,13 +620,14 @@ var _ = Describe("[rfe_id:273][crit:high][vendor:cnv-qe@redhat.com][level:compon
 				}, 60, 1).Should(
 					ContainSubstring(
-						fmt.Sprintf("device socket file for device %s was removed, kubelet probably restarted.", "kvm"),
+						fmt.Sprintf("device socket file for device \\\"%s\\\" was removed, kubelet probably restarted.", deviceResourceName),
```
Use `%q`

0xFelix commented on pkg/virt-handler/device-manager/socket_device.go:
```diff
@@ -134,13 +64,14 @@ func (dpi *SocketDevicePlugin) setSocketPermissions() error {
 
 func (dpi *SocketDevicePlugin) setSocketDirectoryPermissions() error {
-	dir, err := safepath.JoinAndResolveWithRelativeRoot("/", dpi.socketDir)
+	socketDir := path.Dir(dpi.devicePath)
```
Use `filepath` everywhere?

Aseeef commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -149,6 +338,95 @@ func (dpi *DevicePluginBase) setInitialized(initialized bool) {
+	logger := log.DefaultLogger()
+	if dpi.mutateHealthUpdate != nil {
+		var err error
+		err = dpi.mutateHealthUpdate(deviceID, absoluteDevicePath, &healthy)
```
I don't follow. What do you mean?

Aseeef replied:
BTW I also added a unit test for this under usb_device

0xFelix replied:
IIUC it will recompute the state for a given object only when that object triggers this code. The map could contain a cached Unhealthy but state of group could have been Healthy in the meantime. The next time this given object really becomes Unhealthy, it won't trigger the group update since Unhealthy was already cached?

Aseeef commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -210,6 +337,88 @@ func (dpi *DevicePluginBase) setInitialized(initialized bool) {
+		logger.Reason(err).Warningf("failed to get relative path for device %s", absoluteDevicePath)
+		return false
+	}
+	// We derive relDevicePath to be able to enforce containment with JoinAndResolveWithRelativeRoot
```
relDevicePath is short for relativeDevicePath. This wasn't a typo but if you prefer I can make it relativeDevicePath.

0xFelix commented on pkg/virt-handler/device-manager/device_plugin_base.go:
```diff
@@ -84,6 +171,117 @@ func (dpi *DevicePluginBase) ListAndWatch(_ *pluginapi.Empty, s pluginapi.Device
+			logger.Reason(err).Errorf("error watching devices and device plugin directory")
+		case event := <-watcher.Events:
+			logger.V(4).Infof("health Event: %v", event)
+			if event.Name == deviceDirPath && event.Op == fsnotify.Create {
```
I think that could work.

Aseeef commented:
/unhold
/retest

Aseeef committed:
Fixes device-plugin re-registration not working

When kubelet restarts it removes the device plugin socket. The device plugin process never detects the removal and never triggers re-registration with the new kubelet instance. Devices such as vhost-user NICs and SR-IOV VFs become permanently orphaned — pods requesting those devices on the affected node silently fail to schedule, and the only recovery is a manual virt-handler restart.

Signed-off-by: Aseef <contact@aseef.dev>
Co-authored-by: RITANKAR SAHA <ritankar.saha786@gmail.com>

Aseeef commented:
/test

Aseeef commented:
/retest

Aseeef commented:
/remove-lifecycle stale

Aseeef commented:
/remove-lifecycle stale
